pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -269,7 +269,7 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 		runningContainers := false
 		for _, runningPod := range runningPods {
 			if runningPod.ID == volumeToMount.Pod.UID {
-				if len(runningPod.Containers) > 0 {
+				if len(runningPod.Containers) > 0 || len(runningPod.Sandboxes) > 0 {
```

Comment:
Could you add a comment about this change..

So the reasoning here is 
Even if runningPod.Containers == 0, kubelet might still in the process of starting a new container after this point (and mistakenly thought volume is still mounted)

With additional condition runningPod.Sandboxes == 0 which means pod sandboxes were cleaned up (sandbox will be cleaned up after containers were deleted). This can help reduce the chance that after volume is removed from desired state and actual state, new containers might come up..

I am not sure whether this logic can completely avoid the race discovered in this issue, but it should be rather rare. 
cc @Random-Liu

Reply:
In #99639 we avoid using running sandboxes as a proxy for running pods/containers because a pod may be pending with a running sandbox and no running containers.

Furthermore, there's only one sandbox for the pod, so I don't think we should check this for every container, I think it should be checked separately as a one-off.

Reply:
Here we want to make sure no containers are running, also no container in the process of creating. 

Before checking here, we check pod status and container status. But it also might not cover some corner cases.  There was an issue opened https://github.com/kubernetes/kubernetes/issues/69831 
I am now thinking what is the container status in this case, len(runningPod.Containers)=0 len(runningPod.Sandboxes)>0?

Reply:
Why do we need to ensure that runningPod.Containers and runningPod.Sandboxes are both empty at the same time?
1. dswp.isPodTerminated(pod) judges whether the pod has been isPodTerminated base on podStatus
```
	// IsPodTerminated checks if pod is terminated
	func IsPodTerminated(pod *v1.Pod, podStatus v1.PodStatus) bool {
	return podStatus.Phase == v1.PodFailed || podStatus.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(podStatus.ContainerStatuses))
	}
```
If a container is being creating, podStatus.ContainerStatuses probably does not yet have the status of the newly creating container. In some scenarios, such as deleting the pod during the pod is creating, dswp.isPodTerminated(pod) will return an incorrect value.

2. runningPods are obtained from docker:
```
	runningPods, getPodsErr = dswp.kubeContainerRuntime.GetPods(false)
```
dswp.kubeContainerRuntime.GetPods(false) will only return the containers whose state are in running state.
The states of a container are as follows: created, running, paused, stopped

![image](https://user-images.githubusercontent.com/1468240/109744543-738d6f00-7c0d-11eb-9ffe-69ff53b350ab.png)

kubelet creates a container in two steps:
	1. Create a container through runtimeService.CreateContainer, and then the container is in the created state,
	`containerID, err := m.runtimeService.CreateContainer(podSandboxID, containerConfig, podSandboxConfig)`
	2. Start the container through m.runtimeService.StartContainer, and then the container is in the running state
	`err = m.runtimeService.StartContainer(containerID)`
	dswp.kubeContainerRuntime.GetPods(false) cannot get the container in the created state, so len(runningPod.Containers) = 0 does not guarantee that the pod does not have a running container.

3. VolumeManager and syncPod are run in two goroutine. If syncPod is in the process of creating a container, and syncPod has not called runtimeService.CreateContainer to create a container, runningPod.Containers will not include the container whith will be created soon.

Why can't we just use runningPod.Sandboxes to judge the status of the pod?
If the sandbox exits abnormally and the containers are still running (although the probability of this assumption is very low, it may happen that the sandbox container is lost in our cluster). The running container is still writing data to the volume, and the volume unmounting will fail.

Reply:
/hold

Have we resolved this discussion? @ehashman what we're looking for is a signal that containers are in the process of being created. Is `pod.Sandboxes` a good signal of that? Is it possible that a pod is in the process of being created but there are no sandboxes or containers?